### PR TITLE
[SPARK-8888][SQL] Use java.util.HashMap in DynamicPartitionWriterContainer.

### PR DESCRIPTION
**Pull request by rxin**
Just a baby step towards making it more efficient.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala [hunk 1]
```diff
@@ -497,13 +501,14 @@ private[sql] class DynamicPartitionWriterContainer(
...
+  @transient protected var outputWriters: java.util.HashMap[String, OutputWriter] = _
 
   override protected def initWriters(): Unit = {
-    outputWriters = mutable.Map.empty[String, OutputWriter]
+    outputWriters = new java.util.HashMap[String, OutputWriter]
   }
 
   override def outputWriterForRow(row: Row): OutputWriter = {
+    // TODO (SPARK-8888): zip and all the stuff happening here is very inefficient.
     val partitionPath = partitionColumns.zip(row.toSeq).map { case (col, rawValue) =>
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala]: here I think a better solution is to actually project the partition column using expressions, which can also use codegen.

Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala]: Or we can make it as a formatter.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala]: formatter?

**Comment by liancheng** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala (hunk 1 above)
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/commands.scala]: Sorry, for the confusing, I mean just like the `printf`, as the partition column names is fixed.
```
"/date=%s/hour=%d/min=%d".format(...)
```

**Pull request closed without merging**